README.md
# HNT Miner Monitor

![Grafana style for your miner](docs/images/bobcat_blue-by_Wrath.png)

![HNT: 13Vazr2mTQSbu2wBGAkqpaLvJQEdSv5aMd3qpdXFJSw2pfNpqC4](docs/images/wallet.png)

HNT: 13Vazr2mTQSbu2wBGAkqpaLvJQEdSv5aMd3qpdXFJSw2pfNpqC4

## Overview

This repo is used to produce metrics from miner api's and the Helium blockchain. We can use the metrics to diagnose, alert, and prevent poor mining performance. Currently we extract the following metrics:

**Hotspot API**

- activity metrics
- block height
- reward counts (15 minute, 1 hour, 1 day, 7 days, 30 days)
- reward scale
- witnessed
- witnesses

**Bobcat API**

- block height
- cpu temperatures
- height gap
- miner height
- sync status

**LongAP**

- block height
- height gap
- miner connected
- miner height
- online status
- sync status

**Nebra**

- block height
- bluetooth connected
- frequency
- height gap
- lora status
- miner connected
- miner height
- relayed
- sync percent
- sync status

**Sensecap API**

- antenna gain
- block height
- cpu temperature
- cpu used
- dialable
- height gap
- is healthy
- memory total
- memory used
- nat type
- miner connected
- miner height
- relayed
- sd total
- sd used
- sync status
<br />

## System Requirements

Minimum system resources to run this stack are. This will support 5 miners with decent activity

* CoreI5
* 6GB RAM
* 7200 RPM Disk
* x86 / x64 systems (no ARM support)
<br />

## Prerequisite

### Linux +(MacOS): Without using docker ###

If you plan to not use docker to manage your metric collection, you will need to have your monitoring platform setup by using `prometheus`, `prometheus push gateway`, and `grafana`. The pushgateway, from prometheus, will allow us to push metrics to prometheus instead of trying to host the metrics ourselves on an http endpoint.

- **Prometheus**
  - [prometheus push gateway](https://github.com/prometheus/pushgateway)
  - [prometheus](https://prometheus.io/docs/prometheus/latest/installation)

- **Grafana**
  - [grafana](https://grafana.com/docs/grafana/latest/installation)

- **JQ**
  - [jq](https://stedolan.github.io/jq/download/)
<br />

### Linux +(MacOS) & Windows: Using docker and docker-compose ###

#### Linux +(MacOS) ####

You can install `docker` and `docker-compose` by running the `hnt_monitor.sh` script. This will install all of the dependencies necessary, walk you through a setup wizard, and deploy the service. With docker we can create the entire monitoring stack using `docker-compose`.

```bash
$> ./hnt_monitor.sh
```

If you want to only install the dependencies and not setup and deploy, you can pass in the argument `prereq` to the `hnt_monitor.sh` script.

```bash
$> ./hnt_monitor.sh prereq
```
<br />

#### Windows ####

You will need to install `docker` and `docker-compose` manually. You can do that by installing docker desktop for windows.

- [docker](https://docs.docker.com/desktop/windows/install/)
<br />

## Quick Start: Linux +(MacOS) only ##

### Linux +(MacOS): setup & installation ###

The `hnt_monitor.sh` script will install the necessary software and walk you through setting up the monitor stack. Follow the prompts, input your miners, and let the script install the platform

```bash
$> ./hnt_monitor.sh
```
<br />

## Deep Dive ##

### Linux +(MacOS): Without docker ###

Navitage to the `src/conf/` directory and add your miners hotspot address (the public address assigned to the miner on the blockchain "112nVe...") to the `address.list` file. Next, update the `hnt_monitor.conf` file with your prometheus push gateway host and port, <your_miner>_monitor=true, and supply your miners addresses. Then run the hnt monitor script manually in the `src/bin` directory. You can visit the `host:port` of the machine running prometheus pushgaeway and see the new metrics.

```bash
$> ./src/bin/hnt_monitor
```

Run in the background if you want to make it a service

```bash
$> ./src/bin/hnt_monitor &
```

---
**NOTE**

- `*_addresses=`&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;is always referring to the hotspot public address found on the blockchain and explorer. This is the long character string like `112vGthen34DVHjnj...`

- `*_ips=`&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;is always referring to the hotspot private ip addresses that is assigned by your internal router. `192.168.2.1`

- `*_serial_numbers`&nbsp;&nbsp;is always referring to the hotspot unique id identifier that your vendor has supplied

You can add multiple miners by quoting and separating with a space. `"address1 address2 address3"` 

---
<br />

### Linux +(MacOS) & Windows: docker without docker-compose ###

Run the hnt monitor script standalone

```bash
$> docker build -t hnt_monitor -f build/docker/Dockerfile .
$> docker run --rm -it hnt_monitor help     # help menu
$> docker run -d -e HNT_HOTSPOT_MONITOR=true -e HNT_HOTSPOT_ADDRESSES="12345..." -e HNT_PROMETHEUS_PG_HOST=http://my.prometheus-pushgateway.host:9091 hnt_monitor  # Enable hotspot monitoring from helium api
```
<br />

### Linux +(MacOS) & Windows: docker with docker-compose ###

#### Linux +(MacOS): Automated Setup & Deploy ####

Run the `hnt_monitor.sh` script to configure the docker-compose settings and launch the stack.

```bash
$> ./hnt_monitor.sh
```
<br />

#### Linux +(MacOS) & Windows: Manual Setup & Deploy ####

Edit the `hnt_monitor.yml` file and add your miner information to the `hnt_monitor` service environment variables.

```bash
  hnt_monitor:
    container_name: hnt_monitor
    image: hnt_monitor:latest
    build:
      dockerfile: ./build/docker/Dockerfile
      context: .
    environment:
      DO_NOT_REMOVE: "setup"
      HNT_HOTSPOT_MONITOR: "true"
      HNT_HOTSPOT_ADDRESSES: "<myminersaddress> "   # Update your miner address on this line before launching the stack
      HNT_PROMETHEUS_PG_HOST: "http://prometheus_pushgateway:9091"
      HNT_DEBUG: "true"
    networks:
      hnt_monitor:
        ipv4_address: 10.30.0.05
    depends_on:
      - prometheus_pushgateway
```

Check the variables table below for more options that the `hnt_monitor` supports. Once you're satisfied, you can run the `docker-compose`, `up` command, below to launch the stack.

```
$> docker-compose -f hnt_monitor.yml up -d --build
```

Once `docker-compose` completes, you can verify the endpoints in your browser. Open your favorite web browser and check the following endpoints

| Application | Endpoint |
|:-----------:|----------|
| grafana | http://localhost:3000 |
| prometheus | http://localhost:9090 |
| prometheus pushgateway | http://localhost:9091 |
<br />

## Upgrading

### Linux +(MacOS): Automated Upgrade ###

Using the `hnt_monitor.sh` you can provide the `update` command to pull down the latest release and deploy. This will automatically reset any local changes you have made so be sure to copy your work to another directory if you have done your own development. You **do not** need to worry about the `hnt_monitor.yml` changess that exist. The `hnt_monitor.sh` script will use its own .yml file

```bash
$> ./hnt_monitor.sh update
```
<br />

### Linux +(MacOS) & Windows: Manual updates ###

If you have to update manually there are a couple of things you need to do so you dont lose your settings. Make sure to do the following to update your version safely.

- Copy the existing `hnt_monitor.yml` to another directory
- Git pull the repo down and checkout the release version or download and unzip the release
- Copy your `hnt_monitor.yml` to the new release
- Then run `docker-compose -f hnt_monitor.yml up -d --build`

<br />

## Verify Installtation

Check the prometheus push gateway to see metrics have been pushed from the `hnt_monitor`. This service is listening 9091, navigate to `http://localhost:9091` in your browser. You should see a screen like below, with all of the available metrics from the miner collector.

---
**NOTE**

Initial etl processing takes up to 2 minutes to push data. Please allow time for the metrics to be pushed before verification. 

---

![prometheuspg](docs/images/prometheus-pg.png)

### Docker Logs ###

```bash
$> docker logs -f hnt_monitor
```
<br />

## Help

```bash
$> docker run -it --rm hnt_monitor help
```
<br />

## Variables

| Name | Default | Description | Required |
|:----:|---------|-------------|----------|
| `HNT_BLOCKS_URL` | `api.helium.io/v1/blocks` | The helium blocks api url. | `no` |
| `HNT_BOBCAT_IPS` | | If bobcat monitoring enabled, list of ips. Ex: '192.x.x.2 192.x.x.3 192.x.x.etc' | `no` |
| `HNT_BOBCAT_MONITOR` | `false` | Enable or disable bobcat monitoring. Boolean: `(true or false)` | `no` |
| `HNT_DEBUG` | `false` | Turn on debug logging. Boolean: `(true or false)` | `no` |
| `HNT_HELIUM_MONITOR` | `true` | Enable or disable helium monitoring. Boolean: `(true or false)` | `no` |
| `HNT_HOTSPOT_ADDRESSES` | | Hotspot miner addresses to get metrics from. Ex: 'address1 address2 address3 etc' | `no` |
| `HNT_HOTSPOT_MONITOR` | `false` | Enable hotspot monitoring from helium api. Boolean: `(true or false)` | `no` |
| `HNT_HOTSPOT_URL` | `api.helium.io/v1/hotspots` | The helium hotspot api url. | `no` |
| `HNT_LOGFILE` | `stdout` | Send logs to this file | `no` |
| `HNT_LOGPATH` | `/dev/` | Send logs to this path | `no` |
| `HNT_LONGAP_ADDRESSES` | | If longap monitoring enabled, list of ips. Ex: 'address1 address2 address3' | `no` |
| `HNT_LONGAP_MONITOR` | `false` | Enable or disable bobcat monitoring. Boolean: `(true or false)` | `no` |
| `HNT_NEBRA_IPS` | | If nebra monitoring enabled, list of ips. Ex: '192.x.x.2 192.x.x.3 192.x.x.etc' | `no` |
| `HNT_NEBRA_MONITOR` | `false` | Enable or disable nebra monitoring. Boolean: `(true or false)` | `no` |
| `HNT_PROJECT` | `hnt_monitor` | The name of the metric prefix when sending to prometheus. | `no` |
| `HNT_PROMETHEUS_PG_HOST` | `http://localhost:9091` | The prometheus push gateway hostname. | `yes` |
| `HNT_SENSECAP_API_KEY` | | Api key for sensecap | `no` |
| `HNT_SENSECAP_MONITOR` | `false` | Enable or disable sensecap monitoring. Boolean: `(true or false)` | `no` |
| `HNT_SENSECAP_SERIAL_NUMBERS` | | If sensecap monitoring enabled, list of Serial numbers of the sensecap miners | `no` |
| `HNT_TRACE` | `false` | Turn on trace logging. Produces more logs than debug. Boolean: `(true or false)` | `no` |
<br />

## What's next

### [Scheduling](docs/scheduling.md)
### [Setup Grafana](docs/setup.md)
### [Create Dashboards](docs/dashboards.md)
### [Create Alerts](docs/alerts.md)
### [Advanced Settings](docs/advanced.md)
### [Hosted Solution](docs/hosted.md)

<br />

## Tips & Donations

Always welcomed, never required =)

![HNT: 13Vazr2mTQSbu2wBGAkqpaLvJQEdSv5aMd3qpdXFJSw2pfNpqC4](docs/images/wallet.png)

HNT: 13Vazr2mTQSbu2wBGAkqpaLvJQEdSv5aMd3qpdXFJSw2pfNpqC4



HISTORY.md
vNext

# 7.10.0

- **BugFix** - rewards earned bug fixed
- **Enhancement** - arm supported

# 7.9.2

- **BugFix** - nebra collector url remove `/` between uri path and args

# 7.9.1

- **HotFix** - list a single browser agent

# 7.9.0

- **Enhancement** - update miner collectors 
- **HotFix** - collection is broken without user agents

# 7.8.0

- **Enhancement** - update to send all log counts per event

# 7.7.0

- **Enhancement** - update to send warnings per event

# 7.6.0

- **Enhancement** - add totals to log counts
- **Enhancement** - add container log version
- **Enhancement** - log counts reset to 0 on each run

# 7.5.0

- **Enhancement** - add log counts to system metrics
- **Enhancement** - add option to send hnt logs to elasticsearch

# 7.4.2

- **BugFix** - block height missing prometheus gateway

# 7.4.1

- **BugFix** - jobs break when supplying multiple prometheus endpoints

# 7.4.0

- **Enhancement** - improve warning message output to include hotspot address
- **Enhancement** - update READEME to include hosted pricing and support
- **Enhancement** - addresses only update on collector startup
- **Enhancement** - get collector configs from account index
- **Enhancement** - get addresses at the master scheduler

# 7.3.0

- **Enhancement** - increase collector frequency on rewards and activity
- **Enhancement** - add account balance to users backend
- **Enhancement** - move billing index to its own
- **Enhancement** - limit billing balance minimum to 0 instead of negative

# 7.2.0

- **Enhancement** - silence warnings for beacon witness counts
- **Enhancement** - allow mulitple payment wallets for billing analysis
- **BugFix** - billing wallet is not properly applying based on cost
- **BugFix** - miner collectors could die and leave locks forcing container restart

# 7.1.0

- **Enhancement** - give inidividual control to activity etl types

# 7.0.0

- **Enhancement** - witnesses and witnessed directly from activity

# 6.3.0

- **Enhancement** - add miner response time
- **Enhancement** - allow user controlled account share in production

# 6.2.0

- **Enhancement** - update backfill days from 30 to 5
- **Enhancement** - update elasticsearch queries to be more efficient
- **Enhancement** - update min threshold on bytes received 

# 6.1.0 

- **Enhancement** - update increased wait time interval for master schedule
- **Enhancement** - update master schedule is now configurable via env variables

# 6.0.0

- **Enhancement** - update bobcat blue dashboard json
- **Enhancement** - add nginx for api authentication option
- **Enhancement** - add support for api authentication on metric payload 
- **Enhancement** - add support for elasticsearch data storage
- **Enhancement** - add support for multiple relayed port verification
- **Enhancement** - add miner type as a tag for group filtering
- **Enhancement** - add billing etl job
- **Enhancement** - add advanced setup readme

# 5.3.0

- **Enhancement** - updating test for version output
- **BugFix** - metrics were broken in multiple vendor miner setup

# 5.2.0

- **Enhancement** - update bobcat default dashboard json
- **Enhancement** - update bobcat blue dashboard json
- **Enhancement** - update README with latest changes
- **Enhancement** - add version to initialize line of log output
- **Enhancement** - add data transfer metrics to activity count
- **BugFix** - relayed and online status was not properly displayed

# 5.1.0

- **Enhancement** - add online status for all miners
- **Enhancement** - add relayed status for all miners
- **Enhancement** - add user controls for etl jobs
- **Enhancement** - support multiple prometheus endpoints to send data
- **Enhancement** - update dashboads with the shared miner metric names
- **Enhancement** - consolidate jq queries where possible

# 5.0.1

- **BugFix** - activity and reward collections were failing to update failure thresholds

# 5.0.0

- **Enhancement** - update metric names to no longer be unique
- **Enhancement** - add timeout threshold to cursor traversing

# 4.3.0

- **Enhancement** - add bobcat blue dashboard json to extras
- **Enhancement** - add default sensecap dashboard json
- **Enhancement** - sync status threshold updated to 20 from 50
- **Enhancement** - update `hnt_monitor.sh` update command to pull stable release versions
- **Enhancement** - update README
- **Enhancement** - update gauge help message for relayed status
- **Enhancement** - report rewards earned 15m, 1h, 1d, 7d, 30d metrics
- **Enhancement** - collect and report activity metrics

# 4.2.0

- **Enhancement** - update hnt_monitor.sh to not allow blank values
- **Enhancement** - update hnt_monitor.sh to catch existing values and prompt for action
- **Enhancement** - update hnt_monitor.sh with better wording during setup
- **Enhancement** - update hnt_monitor.sh with the `deploy` command
- **BugFix** - sensecap serial numbers variable was incorrect in hnt_monitor.sh

# 4.1.0

- **Enhancement** - update the hnt_monitor.sh script user experience for sensecap
- **Enhancement** - update the reademe, fixed typos
- **Enhancement** - setup.sh script moved to hnt_monitor.sh
- **BugFix** - setup.sh sensecap api key was not properly added

# 4.0.2
- **BugFix** - setup.sh was upgrade command was not working

# 4.0.1
- **BugFix** - setup.sh was not setting hotspot_monitor to true

# 4.0.0

- **Enhancement** - add default dashboard json
- **Enhancement** - add setup.sh to automate install and upgrades
- **Enhancement** - add calculated sync status for all miners
- **Enhancement** - add calculated height gap for all miners
- **Enhancement** - add build test cases
- **Enhancement** - add support for nebra miner
- **Enhancement** - add support for longap miner
- **Enhancement** - add [docker hub](https://hub.docker.com/repository/docker/nhatfield/hnt_monitor) image
- **Enhancement** - sensecap collection updated to collector and etl separation
- **BugFix** - witness etl data not pushed to grafana

# 3.3.1


hnt_monitor.sh
#!/usr/bin/env bash

set -euo pipefail


SOURCE="${BASH_SOURCE[0]}"
while [ -h "$SOURCE" ]; do # resolve $SOURCE until the file is no longer a symlink
  DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
  SOURCE="$(readlink "$SOURCE")"
  [[ $SOURCE != /* ]] && SOURCE="$DIR/$SOURCE" # if $SOURCE was a relative symlink, we need to resolve it relative to the path where the symlink file was located
done
DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
cd ${DIR}

OPT=${1:-"intro"}
HNT_HOTSPOT_ADDRESSES=${HNT_HOTSPOT_ADDRESSES:-""}
YAML=${YAML:-".hnt_monitor.yml"}

if [ ! -f "${YAML}" ]; then
  cp hnt_monitor.yml "${YAML}"
fi

os=$(uname)

if [ "${os}" == "Darwin" ]; then
  sedi="gsed -i"
else
  sedi="sed -i"
fi

intro() {
  clear
  version
  echo
  echo "What do you want to do?"
  echo
  echo "1. Install   [ Install dependencies, setup, and deploy the full monitoring stack ]"
  echo "2. Settings  [ Update miner addresses, serial numbers, ips, etc ]"
  echo "3. Upgrade   [ Upgrade an existing setup to the latest stable release ]"
  echo "4. Deploy    [ Deploy the monitoring stack only ]"
  echo "5. View      [ View the current configuration ]"
  echo "6. Donate    [ Display the HNT wallet backing this project ]"
  echo "7. Exit      [ Exit the program ]"
  echo
  echo
  read title
  
  while [ ! "${title}" ]; do
    echo "Please choose a valid option or cancel the script [CTRL+C]"
    read title
  done

  case ${title} in 
                install|Install|INSTALL|1)
                  clear
                  echo "Starting installation ..."
                  prereq
                  setup
                  deploy
                  ;;
                setting|settings|Setting|Settings|SETTING|SETTINGS|2)
                  setup
                  config
                  ;;
                update|Update|UPDATE|upgrade|Upgrade|UPGRADE|3)
                  update
                  ;;
                deploy|Deploy|DEPLOY|4)
                  deploy
                  ;;
                view|View|VIEW|5)
                  view
                  intro
                  ;;
                donate|Donate|DONATE|6)
                  donate
                  ;;
                exit|Exit|EXIT|7)
                  exit 0
                  ;;
                *)
                  echo "invalid option: ${title}"
                  intro
                  exit 0
                  ;;
  esac
}

view() {
  clear
  grep "HNT_" "${YAML}" | sed 's%^      %%'
  echo
  echo
  echo "Press [enter] key to continue ..."
  read resp

} 

prereq() {
  echo "Verifying software dependencies are installed"
  if [ ! "$(which docker 2>/dev/null)" ] || [ ! "$(which docker-compose 2>/dev/null)" ]; then
    echo "Docker was not detected. Installing..."
    if [ "$(which yum 2>/dev/null)" ]; then
      echo "The script wants to update your system [yum update -y]. This will update all of your systems software! Would you like to proceed? [yes|no]"
      echo
      read up_resp

      case ${up_resp} in
                      y|Y|yes|Yes|YES)
                        yum update -y
                        yum install -y epel-release
                        yum update -y
                        ;;
                      n|N|no|No|NO)
                        echo "skipping updates"
                        ;;
                      *)
                        echo "Please choose a valid option: [yes|no]"
                        ;;
      esac

      yum install docker docker-compose -y
      service docker start || systemctl start docker
    elif [ "$(which apt-get 2>/dev/null)" ]; then
      echo "The script wants to update your system [apt-get update -y] and [apt-get upgrade -y]. This will update all of your systems software! Would you like to proceed? [yes|no]"
      echo
      read up_resp

      case ${up_resp} in
                      y|Y|yes|Yes|YES)
                        apt-get update -y
                        apt-get upgrade -y
                        ;;
                      n|N|no|No|NO)
                        echo "skipping updates"
                        ;;
                      *)
                        echo "Please choose a valid option: [yes|no]"
                        ;;
      esac

      apt-get install docker docker-compose -y
      service docker start || systemctl start docker
    else
      if [ ! "$(which brew 2>/dev/null)" ]; then
        /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
      fi
  
      if [ ! "$(which cask 2>/dev/null)" ]; then
        brew install cask
      fi

      fail=""
      brew install --cask docker-toolbox || fail=true

      if [ "${fail}" == "true" ]; then
        echo -e "\n\nCould not install the docker-toolbox. Check your System & Privacy settings under System Preferences and allow the blocked app \"Oracle VirtualBox\". Then re-run the setup again"
        exit 1
      fi

      brew install docker-machine || brew link --overwrite docker-machine
      docker-machine create --driver "virtualbox" devbox || true
      docker-machine start devbox
      eval "$(docker-machine env devbox)"

      if [ ! "$(grep 'docker-machine env devbox' "${HOME}"/.zshrc 2>/dev/null)" ]; then
        echo 'if ! docker-machine status devbox | grep -q "Running"; then' >> "${HOME}"/.zshrc
        echo '  docker-machine start devbox' >> "${HOME}"/.zshrc
        echo 'fi' >> "${HOME}"/.zshrc
        echo 'eval "$(docker-machine env devbox)"' >> "${HOME}"/.zshrc
      fi

      if [ ! "$(which gsed 2>/dev/null)" ]; then
        brew install gnu-sed
      fi
    fi
  fi
}

setup() {
  clear
  echo "Which miner do you want to add?"
  echo
  echo "1. Bobcat"
  echo "2. LongAP"
  echo "3. Nebra"
  echo "4. Sensecap"
  echo "5. Deploy"
  echo "6. back"
  echo "7. exit"
  echo
  read install

  while [ ! "${install}" ]; do
    echo "Please choose a valid option from the listed items above."
    read install
  done
  
  case ${install} in
                  bobcat|Bobcat|BOBCAT|1)
                    bobcat
                    ;;
                  longap|Longap|LongAP|LONGAP|2)
                    longap
                    ;;
                  nebra|Nebra|NEBRA|3)
                    nebra
                    ;;
                  sensecap|Sensecap|SENSECAP|4)
                    sensecap
                    ;;
                  deploy|Deploy|DEPLOY|5)
                    deploy
                    exit 0
                    ;;
                  back|6)
                    intro
                    exit 0
                    ;;
                  exit|7)
                    exit 0
                    ;;
  esac
}

address_endp() {
  echo "What are the hotspot addresses of the ${id} miner(s)? Please separate them by a [space]"
  echo "ex: 112GezswE 114vTyhUnKo"
  echo
  validate
}

ips_endp() {
  echo "What are the PRIVATE IP addresses of your ${id} miner(s)? Please separate them by a [space]"
  echo "ex: 192.168.0.1 10.10.0.3 172.16.2.5"
  echo
  validate
}

validate() {
  val=$(grep "${endpoints}" ${YAML} 2>/dev/null | sed "s%.*: %%;s%\"%%g") || true
  
  if [ "${val}" ]; then
    echo "Keep existing? [yes|no]"
    echo 
    echo "> \"${val}\""
    echo

    read val_resp

    case ${val_resp} in
                   y|Y|yes|Yes|YES)
                     type=keeping
                     ips=${val}
                     ;;
                   n|N|no|No|NO)
                     type=changing
                     echo "Enter your ips separated by a [space]"
                     echo
                     read ips

                     if [ ! "${ips}" ]; then
                       none
                     fi

                     ;;
                   *)
                     echo "Invalid response: ${val_resp}"
                     echo "[yes|no]"
                     echo "No rchanges committed. Returning to the miner setup menu."
                     sleep 2
                     setup
                     exit 0
                     ;;
    esac
  else
    type=adding
    read ips
    
    if [ ! "${ips}" ]; then
      none
    fi
  fi
}

none() {
  echo "No rchanges committed. Returning to the miner setup menu"
  sleep 2
  setup
  exit 0
}

end_point() {
  echo "${id}: ${type} - ${endpoints} = \"${ips}\""
  if [ "$(grep "${endpoints}" ${YAML} 2>/dev/null)" ]; then
    ${sedi} "s%${endpoints}:.*%${endpoints}: \"${ips}\"%" "${YAML}"
  else
    ${sedi} "s%${monitor}:\(.*\)%${monitor}:\1\n      ${endpoints}: \"${ips}\"%" "${YAML}"
  fi
}

start_point() {
  echo "${id}: enabling - ${monitor} = \"true\""
  if [ "$(grep "${monitor}" ${YAML} 2>/dev/null)" ]; then
    ${sedi} "s%${monitor}:.*%${monitor}: \"true\"%" "${YAML}"
  else
    ${sedi} "s%DO_NOT_REMOVE:\(.*\)%DO_NOT_REMOVE:\1\n      ${monitor}: \"true\"%" "${YAML}"
  fi
}

bobcat() {
  clear
  id=bobcat
  monitor=HNT_BOBCAT_MONITOR
  endpoints=HNT_BOBCAT_IPS

  ips_endp
  start_point
  end_point
  sleep 2
  hotspot
  end  
}

longap() {
  clear
  id=longap
  monitor=HNT_LONGAP_MONITOR
  endpoints=HNT_LONGAP_ADDRESSES

  address_endp
  start_point
  end_point
  sleep 2
  hotspot
  end
}

nebra() {
  clear
  id=nebra
  monitor=HNT_NEBRA_MONITOR
  endpoints=HNT_NEBRA_IPS

  ips_endp
  start_point
  end_point
  sleep 2
  hotspot
  end
}

sensecap() {
  clear
  id=sensecap
  monitor=HNT_SENSECAP_MONITOR
  endpoints=HNT_SENSECAP_SERIAL_NUMBERS

  echo "What are the serial numbers of the ${id} miner(s)? Please separate them by a [space]"
  echo "ex: serial1 serial2"
  echo
  validate

  start_point
  end_point
  sleep 2
  hotspot
  sleep 2

  clear
  id=sensecap
  endpoints=HNT_SENSECAP_API_KEY
  echo "Enter your sensecap api key"
  echo
  validate
  
  end_point
  end
}

hotspot() {
  clear
  id=hotspot
  monitor=HNT_HOTSPOT_MONITOR
  endpoints=HNT_HOTSPOT_ADDRESSES

  address_endp
  start_point

  if [ "${HNT_HOTSPOT_ADDRESSES}" ]; then
    HNT_HOTSPOT_ADDRESSES="${HNT_HOTSPOT_ADDRESSES} ${ips}"
  else
    HNT_HOTSPOT_ADDRESSES="${ips}"
  fi
}

end() {
  echo "Would you like to add another miner? [yes|no]"
  echo
  read end_resp
  
  while [ ! "${end_resp}" ]; do
    echo "Please choose a valid option: [yes|no]"
    read end_resp
  done

  case ${end_resp} in
                   y|Y|yes|Yes|YES)
                     setup
                     ;;
                   n|N|no|No|NO)
                     config
                     ;;
                   *)
                     echo "invalid response: ${end_resp}"
                     end
                     exit 0
  esac    
}

config() {
  HNT_HOTSPOT_ADDRESSES=$(echo "${HNT_HOTSPOT_ADDRESSES}" | tr ' ' '\n' | sort -u | sed 's%^$%%' | tr '\n' ' ')

  if [ "$(grep "HNT_HOTSPOT_ADDRESSES" ${YAML} 2>/dev/null)" ]; then
    ${sedi} "s%HNT_HOTSPOT_ADDRESSES:.*%HNT_HOTSPOT_ADDRESSES: \"${HNT_HOTSPOT_ADDRESSES}\"%" ${YAML}
  else
    ${sedi} "s%DO_NOT_REMOVE:\(.*\)%DO_NOT_REMOVE:\1\n      HNT_HOTSPOT_ADDRESSES: \"${HNT_HOTSPOT_ADDRESSES}\"%" ${YAML}
  fi

  rm -f "${YAML}"\'\'
}

deploy() {
  clear
  echo "Ready to deploy? [yes|no]"
  echo
  read deploy

  while [ ! "${deploy}" ]; do
    echo "Please choose a valid option: [yes|no]"
    read deploy
  done

  case ${deploy} in 
                 y|Y|yes|Yes|YES)
                   docker-compose -f ${YAML} up -d --build
                   finish
                   donate
                   ;;
                 *)
                   echo "Action cancelled"
                   exit 0
                   ;;
  esac
}

finish() {
  clear
  echo "You're all set! Take a look at the docs to setup the application interface"
  echo
  echo "Grafana:                 http://localhost:3000"
  echo "Prometheus Server:       http://localhost:9090"
  echo "Prometheus Push Gateway: http://localhost:9091"
  echo
  echo "Docs:                    https://github.com/nhatfield/hnt_monitor#whats-next"
  echo "Donate:                  HNT: 1359NhpbxJg1jRpDenJvrmD2P3ZN3hWGSGzUF6Uyn828zYdyYVt"
  echo
  echo
  echo "Press [enter] key to continue ..."
  read resp
}

update() {
  git reset --hard
  git pull
  git checkout $(git tag | tail -1 | tr -dc .[:print:].)
  deploy
}

version() {
  echo "HNT Monitor v$(grep '#' HISTORY.md | sed 's%# %%' | sort -V | tail -1)"
}

donate() {
  clear
  cat .donate
  echo "Yes the QR code works. HNT wallet address =)"
  echo "Make your terminal window large enough to see the whole QR. Center your scanner and slowly pull your scanner away from the code until it accepts."
  echo "Thank you =)"
}

case ${OPT} in
            deploy)
              deploy
              ;;
            donate)
              donate
              ;;
            install)
              prereq
              install
              ;;
            intro)
              intro
              ;;
            update|upgrade)
              update
              ;;
            prereq|deps|dependencies)
              prereq
              ;;
            settings|setup)
              setup
              ;;
            view|configs)
              view
              ;;
            version | -v | --version)
              version
              ;;
            *)
              echo "invalid option: ${OPT}"
              ;;
esac
